Change Path into a newtype.

paulp opened the pull request:
When I wrote Path I didn't know that a newtype against
a primitive type allowed you to use it as a receiver.
This removes a whole pointless layer of wrapping. Go go.

paulp commented:
I realize this is diff conflict central, sorry about that.

EECOLOR commented:
As you can see in the travis build you missed one.

EECOLOR commented:
The conflicts are not that bad and I'm happy to solve them.

EECOLOR commented:
I wonder if it's really better to use `x.Str()` vs `string(x)`. According to the specs regular type conversions are free. My guess is that the go compiler is not smart enough to eliminate the method call. I would go for `string(x)`.

EECOLOR commented on suffuse/path.go:
```diff
@@ -39,47 +38,47 @@ func MaybeByteString(result []byte, err error) string {
-func (x Path) OsReadLink() (string, error)                             { return os.Readlink(x.Path)             }
-func (x Path) OsStat() (os.FileInfo, error)                            { return os.Stat(x.Path)                 }
+func (x Path) IsAbs() bool   { return filepath.IsAbs(x.Str()) }
+func (x Path) IsEmpty() bool { return string(x) == ""            }
```
If consistent, this should be `x.Str()`

paulp commented:
The reason I didn't go with `string(x)` is that it's a cast, and I was thinking of the sort of problems shown here: https://gist.github.com/paulp/fca2ce1199abdfb26e69

But maybe that logic only applies with numeric types, where putting a cast into the source paves the way for future failure by abandoning all type safety. If I understand the go rules correctly those `string(x)` calls will stop compiling if we later changed Path to something else. In which case you're right, it should just be string(x).

EECOLOR commented:
From the spec [Conversions](https://golang.org/ref/spec#Conversions):

> Specific rules apply to (non-constant) conversions between numeric types or to and from a string type. > These conversions may change the representation of x and incur a run-time cost. All other conversions only change the type but not the representation of x.

EECOLOR commented:
Relevant part from the spec that relates to your example

> ... It is then truncated to fit in the result type's size. ... The conversion always yields a valid value ...

paulp commented:
I'll fix this up.

paulp commented:
@EECOLOR How about now.

EECOLOR commented:
@paulp Looks good, time ran out for today, so I will merge it tomorrow. And after #20

paulp committed:
Change Path into a newtype.

When I wrote Path I didn't know that a newtype against
a primitive type allowed you to use it as a receiver.
This removes a whole pointless layer of wrapping. Go go.

paulp commented:
I resolved the merge conflicts.

EECOLOR commented on suffuse/path.go:
```diff
@@ -8,30 +8,23 @@ import (
-func NewPath(path string) Path {
-  return Path { path }
-}
-func NewPathRef(path string) *Path {
```
We need this one

EECOLOR replied:
Never mind

EECOLOR commented on suffuse/opts.go:
```diff
@@ -14,7 +14,7 @@ type sfsOpts struct {
 type SfsConfig struct {
   VolName string
-  Config *Path
+  Config Path
```
Ahh, you changed this bit
